src/game-mechanics.js

```diff
@@ -1,59 +1,89 @@
-    [2, 2],
-  ],
-];
+export const makeBoard = (size) =>
+  letIn(makeArrayOf(size), (makeArrayOfSize) =>
+    compose(makeArrayOfSize, makeArrayOfSize)("")
+  );
+
+export const playerHasWin = (player) => (board) =>
+  checkBoardForWin(board) === player;
+
+export const checkBoardForWin = (board) =>
+  compose(checkPlayerStateForWin(board.length), boardToPlayerState);
+
+const checkPlayerStateForWin = (boardSize) => (playerState) =>
+  Object.keys(playerState).reduce(
+    (winner, player) =>
+      checkPlayerState(boardSize)(playerState[player]) ? player : winner,
+    undefined
+  );
 
-export const validateWin = (gameState) => {
-  return winStates.filter((winState) => {
-    console.log(
-      gameState,
-      gameState[winState[0][0]][winState[0][1]] ===
-        gameState[winState[1][0]][winState[1][1]] &&
-        gameState[winState[1][0]][winState[1][1]] ===
-          gameState[winState[2][0]][winState[2][1]]
-    );
+const checkPlayerState = (boardSize) => ({ column, row }) =>
+  letIn({ eqBoardSize: eq(boardSize), gt0: gt(0) }, ({ eqBoardSize, gt0 }) =>
+    letIn(
+      {
+        columnStraight: column.some(eqBoardSize),
+        rowStraight: row.some(eqBoardSize),
+        diagonalStraight: column.all(gt0) && row.all(gt0),
+      },
+      (columnStraight, rowStraight, diagonalStraight) =>
+        columnStraight || rowStraight || diagonalStraight
+    )
+  );
 
-    return (
-      " " !== gameState[winState[0][0]][winState[0][1]] &&
-      gameState[winState[0][0]][winState[0][1]] ===
-        gameState[winState[1][0]][winState[1][1]] &&
-      gameState[winState[1][0]][winState[1][1]] ===
-        gameState[winState[2][0]][winState[2][1]]
-    );
-  });
-};
+/* Player State looks something like this, but can accept any string as a name
+ *
+ * {
+ *  x: { column: [0, 0, 0]
+ *     , row: [0, 0, 0]
+ *     }
+ *  o: { column: [0, 0, 0]
+ *     , row: [0, 0, 0]
+ *     }
+ * }
+ */
+const boardToPlayerState = (board) =>
+  board.reduce(
+    (playerState, row, rowIndex) =>
+      attributeSquareToPlayer(row)(rowIndex)(playerState),
+    {}
+  );
 
-export const hasWon = (gameState) => {
-  return validateWin(gameState).reduce(
-    (result, winState) => result || winState,
-    false
+const attributeSquareToPlayer = (row) => (rowIndex) => (acc) =>
+  row.reduce(
+    (acc, player, columnIndex) =>
+      modifyObject(acc)(player)((playerValue) =>
+        playerValue === undefined
+          ? {
+              columns: modifyArray(
+                columnIndex,
+                add(1),
+                makeArrayOf(row.length - 1)(0)
+              ),
+              rows: modifyArray(
+                rowIndex,
+                add(1),
+                makeArrayOf(row.length - 1)(0)
+              ),
+            }
+          : {
+              columns: modifyArray(columnIndex, add(1), playerValue.column),
+              rows: modifyArray(rowIndex, add(1), playerValue.row),
+            }
+      ),
+    acc
   );
-};
+
```

Comment:
These functions are all curried so we should have something that teaches currying before hand.

Reply:
based on our curriculum for course 1 we do teach currying before hand